runtime/adapters/src/xcm_exchange.rs

```diff
@@ -0,0 +1,123 @@
+	Runtime::Balance: From<u128> + Zero + Into<u128>,
+	DefaultPoolType: Get<PoolType<Runtime::AssetId>>,
+{
+	fn exchange_asset(
+		_origin: Option<&MultiLocation>,
+		give: xcm_executor::Assets,
+		want: &MultiAssets,
+		maximal: bool,
+	) -> Result<xcm_executor::Assets, xcm_executor::Assets> {
+		use orml_utilities::with_transaction_result;
+
+		let account = TempAccount::get();
+		let origin = Runtime::RuntimeOrigin::from(frame_system::RawOrigin::Signed(account.clone()));
+
+		if give.len() != 1 {
+			log::warn!(target: "xcm::exchange-asset", "Only one give asset is supported.");
+			return Err(give);
+		};
+
+		//We assume only one asset wanted as translating into buy and sell is ambigous for multiple want assets
+		if want.len() != 1 {
+			log::warn!(target: "xcm::exchange-asset", "Only one want asset is supported.");
+			return Err(give);
+		};
+		let Some(given) = give.fungible_assets_iter().next() else {
+			return Err(give);
+		};
+
+		let Some(asset_in) = CurrencyIdConvert::convert(given.clone()) else { return Err(give) };
+		let Some(wanted) = want.get(0) else { return Err(give) };
+		let Some(asset_out) = CurrencyIdConvert::convert(wanted.clone()) else { return Err(give) };
+
+		if maximal {
+			// sell
+			let Fungible(amount) = given.fun else { return Err(give) };
+			let Fungible(min_buy_amount) = wanted.fun else { return Err(give) };
+
+			with_transaction_result(|| {
+				Currency::deposit(asset_in, &account, amount.into())?; // mint the incoming tokens
+				pallet_route_executor::Pallet::<Runtime>::sell(
+					origin,
+					asset_in,
+					asset_out,
+					amount.into(),
+					min_buy_amount.into(),
+					vec![Trade {
+						pool: DefaultPoolType::get(),
+						asset_in,
+						asset_out,
+					}],
+				)?;
+				debug_assert!(
+					Currency::free_balance(asset_in, &account) == Runtime::Balance::zero(),
+					"Sell should not leave any of the incoming asset."
+				);
+				let amount_received = Currency::free_balance(asset_out, &account);
+				debug_assert!(
+					amount_received >= min_buy_amount.into(),
+					"Sell should return more than mininum buy amount."
+				);
+				Currency::withdraw(asset_out, &account, amount_received)?; // burn the received tokens
+				Ok(MultiAsset::from((wanted.id, amount_received.into())).into())
+			})
+			.map_err(|_| give)
+		} else {
+			// buy
+			let Fungible(amount) = wanted.fun else { return Err(give) };
+			let Fungible(max_sell_amount) = given.fun else { return Err(give) };
+
+			with_transaction_result(|| {
+				Currency::deposit(asset_in, &account, max_sell_amount.into())?; // mint the incoming tokens
+				pallet_route_executor::Pallet::<Runtime>::buy(
+					origin,
+					asset_in,
+					asset_out,
+					amount.into(),
+					max_sell_amount.into(),
+					vec![Trade {
+						pool: DefaultPoolType::get(),
+						asset_in,
+						asset_out,
+					}],
+				)?;
+				let mut assets = sp_std::vec::Vec::new();
```

Comment:
with_capacity(2) instead  ?

99% of the time, there will be leftover and there will be 2 assets to return.